fmcapi/api_objects.py

```diff
@@ -4906,84 +4906,243 @@ def destination_port(self, action, name=''):
-            fqdns_json = FQDNS(fmc=self.fmc).get()
-            items = ipaddresses_json.get('items', []) + networkgroup_json.get('items', []) + fqdns_json.get('items', [])
-            new_net = None
-            for item in items:
-                if item['name'] == name:
-                    new_net = {'name': item['name'], 'id': item['id'], 'type': item['type']}
-                    break
-            if new_net is None:
-                logging.warning('Network "{}" is not found in FMC.  Cannot add to sourceNetworks.'.format(name))
-            else:
+            if literal != dict():  # This means some value of literal is provided
                 if 'sourceNetworks' in self.__dict__:
-                    duplicate = False
-                    for obj in self.sourceNetworks['objects']:
-                        if obj['name'] == new_net['name']:
-                            duplicate = True
-                            break
-                    if not duplicate:
-                        self.sourceNetworks['objects'].append(new_net)
-                        logging.info('Adding "{}" to sourceNetworks for this ACPRule.'.format(name))
+                    # thus either some literals are already present in sourceNetworks,
+                    # or only objects are present in sourceNetworks
+                    if 'literals' in self.__dict__['sourceNetworks']:
+                        # some literals are already present
+                        duplicate = False
+                        # see if its a duplicate or not. If not, append to the list of
+                        # existing literals in sourceNetworks
+                        for litr in self.sourceNetworks['literals']:
+                            if litr['value'] == literal['value']:
+                                duplicate = True
+                                break
+                        if not duplicate:
+                            self.sourceNetworks['literals'].append(literal)
+                            logging.info('Adding "{}" to sourceNetworks for this ACPRule.'.format(literal))
+                    else:
+                        # this means no literals were present in sourceNetworks,
+                        # and sourceNetworks contains objects only
+                        self.sourceNetworks.update({'literals':[literal]})
+                        # So update the sourceNetworks dict which contained 'objects' key initially
+                        # to have a 'literals' key as well
+                        logging.info('Adding "{}" to sourceNetworks for this ACPRule.'.format(literal))
+                else:
+                    # None of literals or objects are present in sourceNetworks,
+                    # so initialize it with literals and update the provided literal
+                    self.sourceNetworks = {'literals': [literal]}
+                    logging.info('Adding "{}" to sourceNetworks for this ACPRule.'.format(literal))
```

Comment:
Firstly great job getting this done!!

I always hate to get people to do EXTRA work, but I'm thinking that since this is the first significant commit we should take to the time to lay the foundation for our future changes.

For example, the code https://github.com/daxm/fmcapi/blob/78b16a1761c04fd1469491d456a9e0c718f24585/fmcapi/api_objects.py#L4927-L4953 is basically trying to treat a List() like a Dict(). And this code is effectively repeated for each group of changes below.

I think it would be more appropriate to store this as a dictionary which will effectively remove ALL this code and replace it with only a few lines:

```
        if action == 'add':
            if literal != dict():  # This means some value of literal is provided
                self.sourceNetworks['literals'][literal['value']] = literal

```

At the top of source_network (and destination_network) you will need to ensure the attribute has already been built.

```
        if not hasattr(self, 'sourceNetworks'):
            self.sourceNetworks = {'objects': [], 'literals': {}}
```

And lastly in ACPRule.format_data you will need to ensure that 'sourceNetworks' is built not using just self.sourceNetworks but a rendered list of the literals contained in self.sourceNetworks['literals'].

This will significantly reduce the amount of code and improve the performance. Let me know what you guys think.